realtek: phy: multiple cleanups

plappermaul opened the pull request:
This series cleans up old nonsense.

plappermaul committed:
realtek: phy: late phy package patching

Currently phy packages (like RTL8214x/RTL8218x) are patched and
initialized as soon as the first phy of the package is found.
In this situation the shared structure is not finalized because
devm_phy_package_join() has only been called for the first phy.
This is no issue as the patching directly hammers the bus addresses
for the follow-up phys.

In the future we want to simplify the package handling and allow
to access all phy_device structures from only one phy_device of
the package. With this we can use normal phy_read/phy_write.

Switch the probing logic to "late patching". With this we will
initialize the firmware of the package when the last phy of the
package has been found and thus the shared structure is complete.

Provide get_base_phy() as the first package helper that allows
to determine the first phy of the package from any other phy.

While we are here drop the shared structure that only repeats the
phy name and has no other use.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19810
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul committed:
realtek: simplify RTL8214FC handling

Now that we have a get_base_phy() function a lot code of the RTL8214FC
handling can be cleaned up. To name a few:

- use phy_read/phy_write instead of mdiobus functions or the even worse
  phy_package_..._paged() helpers
- replace messages with phydev_info()
- remove if/else statements around copper/fibre handling

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19810
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul committed:
realtek: phy: drop use of mdio package functions

Our phy driver can handle some multiport phys (e.g. RTL8218B
or RTL8214FC). To access arbitrary ports some package access
functions have been defined. These were implemented in the
mdio bus with poor knowledge about the phy/mdio dependencies.
So they add unneeded complexity to the bus and the phy driver
must access these external functions directly.

Provide a new helper get_package_phy() that can derive any
phy device in a package from a given phy of that package.
Make use of this local helper and cut the mdio dependency.

While we refactor several firmware patching functions rename
the loop variables to "port" to better indicate what we are
working on.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19810
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul commented:
@GoetzGoerisch this is close to a full makeover of the RTL838x PHY detection/patching/handling. Would be nice if you have time to test and confirm.

bevanweiss commented on target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -721,41 +727,43 @@ static int rtl8380_configure_int_rtl8218b(struct phy_device *phydev)
+	for (int port = 0; port < 8; port++) {
+		patchphy = get_package_phy(phydev, port);
+		phy_write_paged(patchphy, RTL838X_PAGE_RAW, RTL8XXX_PAGE_SELECT, RTL821X_PAGE_PATCH);
+		phy_write_paged(patchphy, RTL838X_PAGE_RAW, 0x10, 0x0010);
```
Any chance of reducing the magic numbers?

bevanweiss commented on target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -721,41 +727,43 @@ static int rtl8380_configure_int_rtl8218b(struct phy_device *phydev)
-			phydev_err(phydev,
-			           "ERROR: Port %d not ready for patch.\n",
-			           mac + p);
+			phydev_err(patchphy, "not ready for patch.\n");
```
I think it'd still be nice here to keep the Port # in the error message... otherwise an error is going to require kernel hacking to work out which port to further investigate.

plappermaul committed:
realtek: phy: add RTL8214x/RTL8218x patch helper

The patching sequence of the RTL8214x/8218x is very similar.
Especially the preparation for readiness is always the same.
Provide a common helper to improve readability.

While we are here clean up the changed functions

- Sort variable definitions according to upstream
- simplify some messages

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19810
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

bevanweiss commented on target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -3620,86 +3551,49 @@ static int rtl8214fc_phy_probe(struct phy_device *phydev)
-			return rtl8380_configure_ext_rtl8218b(phydev);
-		}
+	devm_phy_package_join(&phydev->mdio.dev, phydev, base_addr, 0);
+	if (phydev->mdio.addr == base_addr + 7) {
```
Is 0 not a valid address for configuration?

systemcrash reviewed (commented):
These where implemented -> These were implemented

systemcrash reviewed (commented):
readyness -> readiness

systemcrash reviewed (commented):
provide -> provides

plappermaul commented:
Load average coming from that series?

GoetzGoerisch commented:
> Load average coming from that series?

Initial observation:
Is already higher with 6.12 then 6.6 and this series increased it even more. Rebuilding your latest changes.

Could you resolve the conflicts with main please?

plappermaul commented:
rebased on master. 

Regarding the load. With only one network plugged in and 4 SFPs inserted (not connected) I see after 10 minutes uptime.

```
root@OpenWrt:~# cat /proc/loadavg
0.00 0.00 0.00 2/42 3638
```

GoetzGoerisch commented:
> rebased on master.

Thank you, currently rebuilding.

running earlier commit:
 ```
root@switch07:~# cat /proc/loadavg 
1.59 1.41 0.86 1/70 2994
 ```

```
root@switch07:~# cat /etc/config/network 

config interface 'loopback'
	option device 'lo'
	option proto 'static'
	list ipaddr '127.0.0.1/8'

config globals 'globals'
	option packet_steering '1'
	option steering_flows '128'

config device 'switch'
	option name 'switch'
	option type 'bridge'
	option macaddr '<REDACTED>'
	list ports 'lan1'
	list ports 'lan2'
	list ports 'lan3'
	list ports 'lan4'
	list ports 'lan5'
	list ports 'lan6'
	list ports 'lan7'
	list ports 'lan8'
	list ports 'lan9'
	list ports 'lan10'
	list ports 'lan11'
	list ports 'lan12'
	list ports 'lan13'
	list ports 'lan14'
	list ports 'lan15'
	list ports 'lan16'
	list ports 'lan17'
	list ports 'lan18'
	list ports 'lan19'
	list ports 'lan20'
	list ports 'lan21'
	list ports 'lan22'
	list ports 'lan23'
	list ports 'lan24'
	list ports 'lan25'
	list ports 'lan26'
	list ports 'lan27'
	list ports 'lan28'

config bridge-vlan 'lan_vlan'
	option device 'switch'
	option vlan '1'
	list ports 'lan1:u*'
	list ports 'lan2'
	list ports 'lan3'
	list ports 'lan4'
	list ports 'lan5'
	list ports 'lan6'
	list ports 'lan7'
	list ports 'lan8'
	list ports 'lan9'
	list ports 'lan10'
	list ports 'lan11'
	list ports 'lan12'
	list ports 'lan13'
	list ports 'lan14'
	list ports 'lan15'
	list ports 'lan16'
	list ports 'lan17'
	list ports 'lan18'
	list ports 'lan19'
	list ports 'lan20'
	list ports 'lan21'
	list ports 'lan22'
	list ports 'lan23:u*'
	list ports 'lan24'
	list ports 'lan25'
	list ports 'lan26'
	list ports 'lan27'

config device
	option name 'switch.1'
	option macaddr '<REDACTED>'

config interface 'lan'
	option device 'switch.10'
	option proto 'static'
	option ip6assign '60'
	list ipaddr '10.1.10.7/24'
	option gateway '10.1.10.1'
	option broadcast '10.1.10.255'

config bridge-vlan
	option device 'switch'
	option vlan '10'
	list ports 'lan28:t'

config bridge-vlan
	option device 'switch'
	option vlan '90'
	list ports 'lan28:t'

config interface 'mgmt6'
	option proto 'dhcpv6'
	option device 'switch.10'
	option reqaddress 'try'
	option reqprefix 'no'
	option norelease '0'

config interface 'mgmt'
	option proto 'dhcp'
	option device 'switch.10'

```

GoetzGoerisch commented:
> Thanks. Regarding loadavg. How many ports connected and which ones (SFP)?

1 device - RJ45 with static IPv4 config, untagged VLAN 1
1 SFP-DAC connected to a VLAN trunk port on another switch, currently unconfigured on this switch.

plappermaul commented:
Cannot reproduce. Can you check user/kernel times of running processes after some uptime

`cat /proc/[0-9]*/stat | awk '{ print $14 " " $15 " " $2  }'
`

idle RTl838x system gives after ~30 minutes

```
root@OpenWrt:/proc/1926# cat /proc/[0-9]*/stat | awk '{ print $14 " " $15 " " $2  }'
18 474 (procd)
0 0 (kworker/R-mm_percpu_wq)
22 45 (rpcd)
0 0 (rcu_tasks_trace_kthread)
0 0 (dropbear)
0 69 (ksoftirqd/0)
0 0 (spi0)
21 283 (netifd)
0 0 (kworker/R-inet_frag_wq)
0 0 (oom_reaper)
1 38 (uhttpd)
0 0 (kworker/R-writeback)
28 0 (kcompactd0)
0 0 (kworker/R-kblockd)
0 0 (watchdogd)
0 0 (kswapd0)
0 12 (ntpd)
6 14 (ntpd)
0 1 (kthreadd)
0 8 (kworker/u4:1-events_unbound)
2 18 (ash)
0 0 (kworker/0:2)
0 2 (cat)
0 0 (kworker/R-kthrotld)
0 50 (kworker/0:3-events_freezable_pwr_efficient)
0 66 (kworker/u4:2-events_unbound)
0 0 (pool_workqueue_release)
0 0 (kworker/R-mld)
0 0 (kworker/R-ipv6_addrconf)
0 0 (kworker/R-dsa_ordered)
0 3 (kworker/u4:4-events_unbound)
0 0 (kworker/R-slub_flushwq)
0 0 (kworker/0:1H)
0 0 (irq/11-keys)
0 0 (kworker/R-netns)
0 1071 (kworker/0:0-events_power_efficient)
0 18 (jffs2_gcd_mtd7)
0 134 (kworker/0:0H-kblockd)
6 18 (ubusd)
1 8 (login)
429 3 (urngd)
0 1120 (kworker/0:1-events_power_efficient)
71 28 (logd)

```

plappermaul commented:
Is this POE and realtek-poe running ?

GoetzGoerisch reviewed (approved):
run-tested: HPE 1920-24G (JG924A)

Thank you!

hauke commented on target/linux/realtek/files-6.12/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -632,20 +637,41 @@ static int rtl8390_configure_generic(struct phy_device *phydev)
+		patchphy = get_package_phy(phydev, port);
+		phy_write_paged(patchphy, RTL821X_PAGE_PATCH, 0x10, 0x10);
+	}
+	msleep(500);
+	for (port = 0; port < maxport; port++) {
+		patchphy = get_package_phy(phydev, port);
+		for (i = 0; i < 100 ; i++) {
+			val = phy_read_paged(patchphy, RTL821X_PAGE_STATE, 0x10);
+			if (val & 0x40)
+				break;
+		}
+		if (i >= 100) {
+			phydev_err(patchphy, "not ready for patch.\n");
+			return -1;
+		}
+	}
```
This looks like you are polling for some value.

Do you rally have to sleep for 500ms or can you also start the polling immediately and wait for 10ms after each unsuccessful check? The you only wait as long as t is needed and still would also catch situations where it needs more than 500ms.
